internal/schema/schema.go

```diff
@@ -0,0 +1,343 @@
+	}
+
+	var inputValidationErrorList *ValidationErrorList
+	ok := errors.As(inputValidationErr, &inputValidationErrorList)
+	if !ok {
+		inputValidationErrorList = nil
+	}
+
+	return MergeValidationErrorLists(inputValidationErrorList, NewValidationErrorList(principalValidation, schemav1.ValidationError_SOURCE_PRINCIPAL))
+}
+
+func (m *Manager) validateResource(ctx context.Context, resourceKind string,
+	resourceAttributes map[string]*structpb.Value) error {
+	m.mu.RLock()
+	defer m.mu.RUnlock()
+
+	resourcePropertiesMap, ok := m.resourceProperties[resourceKind]
+	if !ok {
+		m.log.Warnf("no schema properties found for the kind '%s'", resourceKind)
+		return nil
+	}
+
+	inputValidationErr := m.validateInput(resourceAttributes, resourcePropertiesMap, schemav1.ValidationError_SOURCE_RESOURCE)
+	if inputValidationErr != nil && !IsValidationErrorList(inputValidationErr) {
+		return fmt.Errorf("failed to validate the input: %w", inputValidationErr)
+	}
+
+	resourceAttrBytes, err := json.Marshal(resourceAttributes)
+	if err != nil {
+		return fmt.Errorf("failed to marshal resource attributes: %w", err)
+	}
+
+	resourceSchema, ok := m.resourceSchemas[resourceKind]
+	if !ok {
+		if m.conf.Enforcement == EnforcementReject {
+			return fmt.Errorf("no schema found for the kind '%s'", resourceKind)
+		}
+		m.log.Debugf("no schema found for the kind '%s'", resourceKind)
+		return nil
+	}
+
+	resourceValidation, err := resourceSchema.ValidateBytes(ctx, resourceAttrBytes)
+	if err != nil {
+		return fmt.Errorf("failed to validate resource attributes: %w", err)
+	}
+
+	var inputValidationErrorList *ValidationErrorList
+	ok = errors.As(inputValidationErr, &inputValidationErrorList)
+	if !ok {
+		inputValidationErrorList = nil
+	}
+
+	return MergeValidationErrorLists(inputValidationErrorList, NewValidationErrorList(resourceValidation, schemav1.ValidationError_SOURCE_RESOURCE))
+}
+
+func (m *Manager) walkInputProperties(path string, properties map[string]*structpb.Value, writeTo map[string]string) {
+	for key, value := range properties {
+		structVal, ok := value.Kind.(*structpb.Value_StructValue)
+		if !ok {
+			writeTo[fmt.Sprintf("%s/%s", path, key)] = fmt.Sprintf("%s/%s", path, key)
+			continue
+		}
+		if structVal.StructValue.Fields != nil {
+			m.walkInputProperties(fmt.Sprintf("%s/%s", path, key), structVal.StructValue.Fields, writeTo)
+		}
+	}
+}
+
+func (m *Manager) walkSchemaProperties(path string, properties map[string]*schemav1.JSONSchemaProps,
+	writeTo map[string]string) {
+	for name, value := range properties {
+		if value.Type != "object" {
+			writeTo[fmt.Sprintf("%s/%s", path, name)] = fmt.Sprintf("%s/%s", path, name)
+			continue
+		}
+		if value.Properties != nil {
+			m.walkSchemaProperties(fmt.Sprintf("%s/%s", path, name), value.Properties, writeTo)
+		}
+	}
+}
+
+func (m *Manager) ReadOrUpdateSchemaFromStore(ctx context.Context) error {
+	ctx, span := tracing.StartSpan(ctx, "schema.ReadOrUpdateSchemaFromStore")
+	defer span.End()
```

Comment:
This method is not called from any of the API methods, is it? These trace spans won't be very useful because of that.